### PR TITLE
Context-aware ExactPrint grafting for HsExpr

### LINKED ISSUE #1486: Wingman inserts layout-breaking spaces when case-splitting in do block
as reported by @jhrcek:

```haskell
data Thing = A | B | C

getThing :: IO Thing
getThing = pure A

f :: Int -> IO ()
f i = do
    thing <- getThing
    _
```

this case splits to

```haskell
f :: Int -> IO ()
f i = do
    thing <- getThing
     (case thing of
       A -> _
       B -> _
       C -> _
    )
```

but notice the leading space. In expressions it doesn't matter, but here YIKES.

Seems related to https://github.com/alanz/ghc-exactprint/issues/91, except that we aren't calling `addAnnotationsForPretty` anymore. Maybe generated splices just need a `setPrecedingLine (-1) 0` to get rid of this leading column?

### PR DESCRIPTION
**Pull request by isovector**
In #1486, it's pointed out that always inserting a leading space when grafting trees will break layout (and thus the parse) if you're grafting into a `do` block. Yikes.

This PR adds a function `graftExpr` which tracks the tree context of the graft's destination. If the immediate parent of the destination is a function application, or a few other things, we need to insert that space.

And hey, while we're looking at context, let's use it to determine if we need parentheses!

Fixes #1486

### CODE REVIEW

**Review by jneira — APPROVED** [review]
lgtm

**Comment by isovector** [comment]
Maybe `graft` should be a member of the typeclass, and then we can automatically dispatch to `graftExpr`. Does anyone have a strong opinion on this?

File: ghcide/src/Development/IDE/GHC/ExactPrint.hs [hunk 1]
```diff
@@ -178,30 +179,71 @@ transformM dflags ccs uri f a = runExceptT $
...
+ You want to use 'graftExpr' instead of this function when @ast ~ 'HsExpr'@.
 -}
 graft ::
     forall ast a.
     (Data a, ASTElement ast) =>
     SrcSpan ->
     Located ast ->
     Graft (Either String) a
-graft dst = graftWithoutParentheses dst . maybeParensAST
+graft dst = graft' True dst . maybeParensAST
```
Comment (isovector) [inline ghcide/src/Development/IDE/GHC/ExactPrint.hs:227]: Maybe graft should be a member of the typeclass, and then we can automatically dispatch to graftExpr. Does anyone have a strong opinion on this?

Comment (isovector) [reply ghcide/src/Development/IDE/GHC/ExactPrint.hs:227]: Yeah, that's clearly the right move here.

File: ghcide/src/Development/IDE/GHC/ExactPrint.hs [hunk 2]
```diff
@@ -178,30 +177,57 @@ transformM dflags ccs uri f a = runExceptT $
...
+needsParensSpace ExplicitSum{}   = (All False, All False)
+needsParensSpace HsCase{}        = (All False, All False)
+needsParensSpace HsIf{}          = (All False, All False)
+needsParensSpace HsMultiIf{}     = (All False, All False)
+needsParensSpace HsLet{}         = (All False, All False)
+needsParensSpace HsDo{}          = (All False, All False)
+needsParensSpace ExplicitList{}  = (All False, All False)
+needsParensSpace RecordCon{}     = (All False, All False)
+needsParensSpace RecordUpd{}     = mempty
+needsParensSpace _               = mempty
```
Comment (Ailrun) [inline ghcide/src/Development/IDE/GHC/ExactPrint.hs:208]: Do we have a case that "Needs paren" and "Needs space"" are different?

Comment (isovector) [reply ghcide/src/Development/IDE/GHC/ExactPrint.hs:208]: Today, no, but in general they are not the same. For example, if you are grafting into `a + _`, you need the space but you only need parens if you contain an operator with associativity < 5.

Comment (jneira) [reply ghcide/src/Development/IDE/GHC/ExactPrint.hs:208]: Love the new plugin name :smile:

**Comment by isovector** [comment]
I'm planning on merging this by end of day if nobody disagrees!

**Review by konn — APPROVED** [review]
LGTM 👍 Thank you for nice change, and sorry for late reply!

**Pull request merged**